### PR TITLE
Sparse simulator provides little-endian labels to dump callback

### PR DESCRIPTION
**Pull request by DmitryVasilevsky**
Dump callback requires little-endian labels as it uses BasisStateLabelToBigInt, which assumes data is little-endian. However, labels provided by the Sparse Simulator used to be in big-endian format. This resulted in wrong naming of basis vectors in the dump. Now labels are converted to little-endian format.
This resolves issue https://github.com/microsoft/qsharp-runtime/issues/1081
This change also gets rid of defines in the sparse simulator. Use of defines is not a good practice in general. They also prevented reasonable debugging and searching (they used literal string compositions to name functions).

### CODE REVIEW

File: src/Common/cmake/secure_dependencies.cmake [hunk 1]
```diff
@@ -37,15 +37,21 @@ macro(locate_win32_spectre_static_runtime)
...
         execute_process(
             COMMAND "${_vswhere_tool}" -latest -products * -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -find **/14.*.*/**/lib/spectre/x64 -sort
             OUTPUT_VARIABLE _vs_install_loc_out
             RESULT_VARIABLE _vs_where_exitcode
             OUTPUT_STRIP_TRAILING_WHITESPACE)
+        message(INFO "*** _vs_install_loc_out: ${_vs_install_loc_out}")
+
         file(TO_CMAKE_PATH "${_vs_install_loc_out}" SPECTRE_LIB_PATH_OUT)
+        message(INFO "*** SPECTRE_LIB_PATH_OUT: ${SPECTRE_LIB_PATH_OUT}")
+
```
Comment (kuzminrobin) [inline src/Common/cmake/secure_dependencies.cmake:54]: I assume that the changes in this file will be reverted (or commented out) before merging.

File: src/Simulation/NativeSparseSimulator/CMakeLists.txt [hunk 2]
```diff
@@ -30,6 +30,10 @@ if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
 	target_compile_options(Microsoft.Quantum.SparseSimulator.Runtime PUBLIC -O3 -ftree-vectorize -mavx2 -mfma)
 endif()
 
+if (MSVC)
+    add_compile_options(/bigobj)
+endif()
```
Comment (kuzminrobin) [inline src/Simulation/NativeSparseSimulator/CMakeLists.txt:35]: How does this relate to the issue being fixed?
Code comments wouldn't harm.

**Review by kuzminrobin — APPROVED** [review]
Approving with minor comments.

File: src/Simulation/NativeSparseSimulator/capi.cpp [hunk 3]
```diff
@@ -52,45 +52,56 @@ extern "C"
...
 
-// Generic single-qubit gate
-#define FWDGATE1(G)                                                                                                    \
-    MICROSOFT_QUANTUM_DECL void G##_cpp(simulator_id_type sim_id, logical_qubit_id q)                                                   \
-    {                                                                                                                  \
-        getSimulator(sim_id)->G(q);                                                                  \
+    // single-qubit gates
+    MICROSOFT_QUANTUM_DECL void X_cpp(simulator_id_type sim_id, logical_qubit_id q)
+    {
+        getSimulator(sim_id)->X(q);
```
Comment (kuzminrobin) [inline src/Simulation/NativeSparseSimulator/capi.cpp:58]: I don't insist, but these changes are hardly related to the problem being fixed.

File: src/Common/cmake/secure_dependencies.cmake (hunk 1 above)
Comment (DmitryVasilevsky) [reply src/Common/cmake/secure_dependencies.cmake:54]: Actually, I'm planning to keep them. They help a lot when figuring out what went wrong. I didn't have some components installed and the build was failing in a weird way - such as a function not receiving enough arguments. Turned out variables were not set earlier. By printing them out I was able to see what's wrong.

File: src/Simulation/NativeSparseSimulator/capi.cpp (hunk 3 above)
Comment (DmitryVasilevsky) [reply src/Simulation/NativeSparseSimulator/capi.cpp:58]: Agreed. I just fixed stuff that prevented me from debugging code efficiently. I can move them to a separate PR.

File: src/Simulation/NativeSparseSimulator/CMakeLists.txt (hunk 2 above)
Comment (DmitryVasilevsky) [reply src/Simulation/NativeSparseSimulator/CMakeLists.txt:35]: Normally we build this with clang. Building with cmake/MSVC will fail without this switch.

Comment (kuzminrobin) [reply src/Simulation/NativeSparseSimulator/CMakeLists.txt:35]: Why would one need to build it with MSVC? To debug easier? And just for local builds (not CI)?
If yes then I would recommend to comment this out and provide extra comments explaining why one may need this.

**Comment by cgranade** [comment]
Before releasing this change, please make sure to test that it doesn't modify the dumps shown by IQ# (either in Q# standalone or Q# + Python notebooks). In particular, the case shown by @tcNickolas in #1081 should be checked from a Q# standalone notebook. We may also want to consider either an e2e integration test or iqsharp unit test to make sure that the fix to #1081 does not cause a regression in the use of `%simulate_sparse` / `.simulate_sparse`.

File: src/Simulation/NativeSparseSimulator/CMakeLists.txt (hunk 2 above)
Comment (DmitryVasilevsky) [reply src/Simulation/NativeSparseSimulator/CMakeLists.txt]: Moving to a separate PR.

**Comment by DmitryVasilevsky** [comment]
Since we already forked the code for September release, I'm merging this in. We'll verify all dependencies before the next release.

**Pull request merged**